### ISSUE TITLE
search.js

### ISSUE THREAD

**Issue by lachlan-00** [ampache/ampache#2130]
hey @AshotN can you please look at the lib/javascript/search.js file? 

when searching i get one star because the js isn't selecting the correct star rating.
![image](https://user-images.githubusercontent.com/1305249/68292614-f9f87600-00d7-11ea-874a-7f07fa57f515.png)

I've reverted some changes i made from scrutinizer which caused this last time. Right now this is happening in firefox only as well. so i'm not sure if other changes have affected this yet

**Comment by lachlan-00** [comment]
works again on a different laptop must have been cached.

changes reverted here. i wonder if there's a better way to set these?
https://github.com/ampache/ampache/commit/696586afa5fa22d32234fef92a64f14097dbe53a

**Comment by AshotN** [comment]
`===` should be used instead of `==`

I’ll try this on my computer later and see if I can replicate. The JS for ampache is really old and ugly.

**Comment by lachlan-00** [comment]
thank you! i haven't done a lot of changes to the js as it tends to break a lot of things in the interface

**Comment by AshotN** [comment]
It’s a mess I know. I wonder the prospects of redesigning the front end with Vue or React after the server-side Laravel update.

**Comment by kuzi-moto** [comment]
> It’s a mess I know. I wonder the prospects of redesigning the front end with Vue or React after the server-side Laravel update.

I think Laravel handles the front end as well, unless you're taking about customising the look past that. But until someone is interested in continuing developing for it, efforts will be going to the main branch.

I've been working on cleaning up the frontend JS to move more of the styling to CSS. There is a bit of work but not too bad.

**Comment by AshotN** [comment]
Well Laravel mixes with Vue really well. You can use it to generate the page similar to what’s happening now. But I’m saying using Laravel strictly as back-end and then using JS to fetch. Only downside is JavaScript can’t be disabled. 

If we redo the CSS we should try to use Stylus/LESS/SASS

**Comment by AshotN** [comment]
@lachlan-00 can you give me steps to reproduce your issue?

**Comment by lachlan-00** [comment]
Changing == to ===
And != To !== Will break the rules

**Comment by AshotN** [comment]
Something is wrong then if == works but === doesn't

Relying on == is a dangerous game because it means we aren't sure about our types. This is where TypeScript would've been nice.

**Comment by AshotN** [comment]
Also can I get steps to reproduce, I'm not sure if I'm looking at the correct thing.

**Comment by lachlan-00** [comment]
Steps with changed file

Open any search. (e.g. search.php?type=song)

search for anything star related
![image](https://user-images.githubusercontent.com/1305249/68361957-59ef2b00-0171-11ea-8139-7f5e0ee3b7be.png)

set it above 1 star and search
![image](https://user-images.githubusercontent.com/1305249/68362044-9e7ac680-0171-11ea-8201-2747ee9ac5e9.png)

Results are correct but the search box goes back to 1 star

**Comment by AshotN** [comment]
Oh I see, ya a lot of things seem to reset. The search bar for example does the same thing. I could fix it, but the entire client-side needs a rewrite in my opinion. I would rather discuss if we should consider a massive overhaul. If we can make the API more complete and use something like Vue or React for the client side, I think it would alleviate a lot of the spaghetti and issues we are experiencing across the client-side. 

Frankly it would be a massive overhaul, but since JavaScript is practically required to use a lot of the functionality of Ampache I think it's worth the compromise of making JavaScript required. 

Vue works pretty well with Laravel, if we rather prioritize Laravel first. But most my experience is in React, so I'd rather not use Vue unless someone feels confident enough to audit my Vue code.

A lot of our JavaScript right now is frankly awful, I understand the project is old but I've found things that don't even make sense when you take the age of the project into account. For example, when the JS loads it runs this line which is bizarre`$(".default_hidden").hide();`. All this line does is add `display: none`, so why not just make a CSS class that would do the same thing. 

Another example is the way the alternating row background-color is set. Instead of using CSS `nth-child` which has major support for around 10 years now, there is a bunch of confusing PHP code that sets the class to `even` or `odd`. 

If we keep just patching small issues like this we'll never get anywhere, our client-side is beyond ancient at this point, and even though it will be unpleasant and take a while, I think it needs to be completely redone.

**Comment by kuzi-moto** [comment]
> If we keep just patching small issues like this we'll never get anywhere, our client-side is beyond ancient at this point, and even though it will be unpleasant and take a while, I think it needs to be completely redone.

I agree, the client-side code is really a pretty big mess. There are many best practices that should be implemented, separation of logic and presentation being a big one, and using classes more than ids for second.

There has been a lot of discussion about updating Ampache over the last several years, but there has never been enough of a push to follow through with any of the efforts. Prior to Lachlan taking over, Ampache development had slowed to practically zero. 

In my view, unless there are at least 2-3 dedicated people working on either Laravel or completely re-doing the current code, any attempts will fail. Laravel had been worked on for over 2 years and is missing most of the major functionality. I think an incremental, slow approach is going to be the only way the project will survive, or it will risk everyone burning out.

**Comment by AshotN** [comment]
Problem with the slow incremental approach is ambitions wear thing and then the changes never happen. The Laravel branch is dead, and it's 2 years old. I believe the user interface needs to be completely scraped and redone, with modern issues considered and not 15 year old compatibility layers. @lachlan-00 it's your call what we do, we can either go JavaScript front-end with React or Vue, or use a PHP framework. I started a branch on my fork for React, and the API just needs to be amended to accommodate a new UI.

The API currently, as far as I remember, does not allow logging in to a user account with regular username/password, it works with API keys instead. So that would be an example of something the API would need to move forward.

**Comment by AshotN** [comment]
The API does allow username/password logins, I'm going to see how far I can get with my old JSON implementation.

**Comment by lachlan-00** [comment]
api password implementation is user & sha256(unixtime+sha256(password)) (https://github.com/ampache/ampache/wiki/XML-API)
```
$time = time();
$key = hash('sha256','mypassword');
$passphrase = hash('sha256',$time . $key);
```

TBH my approach is that working on the code and api will make any future work easier but i don't want to throw the code we have out without a fully functional alternative. I think php over JS for format of that solution. Ampache has a big advantage in it's history compared to writing an application from scratch.

My goals here have been to clean up and continue Ampache in it's current form which i'm happy to do and keep doing. i'm also happy to extend the api with a view of dropping the original web interface at some stage.

maybe one day that will lead to a core server that has a separate ui configured over the api but for me Ampache with a web ui + android apps actually isn't that terrible.

continued effort on Subsonic, the xml-api (json in the future) and a renewed focus on supporting interaction with the server should make the ui redundant in the future. hell, an alternative web-ui could just become another ampache app.

**Comment by AshotN** [comment]
I have a working login in example on one my Fork. I’m going to work on a React UI, the API isn’t bad as is but JSON would have to be done first, I’ll implement whatever is missing in my JSON API as needed 

The API is in my opinion the most important thing, clients can be made with it. 

I’ll just see what I can get done but I think I’m not going to replicate the current theme. It’s way to cluttered. But I really think we should sit down and discuss what languages and technologies to use. My current branch is using React, Stylus, Parcel, Babel, and TypeScript. I don’t know if maintainers would prefer to use something else. 

So while we can resolve this issue, I don’t think it’s worth maintaining the current theme.

**Comment by lachlan-00** [comment]
agree that the api is critical

**Comment by kuzi-moto** [comment]
Are any configuration options not exposed in the API? If yes, then Ampache should still have some interface. Doesn't have to be reborn, but at least look nice and modern.

**Comment by AshotN** [comment]
The idea is the allow everything to be controllable from the API. That way clients can be built to replace whatever Ampache ships. It also makes building a client significantly easier, you can follow along on my attempts here https://github.com/AshotN/ampache/tree/ReactClient

**Comment by lachlan-00** [comment]
I think it's a good idea. I'm just running through all the areas that have changed testing things out but i think we'll release soon and then the api will become probably the primary focus outside bug fixing.

Subsonic is still popular because of how solid the api was. i want to do the same things (except not go proprietary) and really make Ampache available as possible for others.

**Comment by kuzi-moto** [comment]
Does it just make sense to scrap the current frontend and then build a simple client to ship with Ampache using the API?

**Comment by wagnered** [comment]
Maybe I'm a little thick headed because of my age, but I understand that the current front end is the web interface. I definitely don't want to see that removed.  Does the API allow full administrative control of Ampache?  If not, please explain how a "simple client", desktop or other, can have full control of the Ampache server.  

I agree that Ampache should have a "matching" client, for either or both desktop and android.  I think that having a "matching" player is one feature that makes plex one of the more favored music streaming servers with Ampache coming in about fourth in several surveys I've seen.

There is a simple android client called Power Ampache which appears to be kept relatively up-to-date.  I'm not able to install it on my broken android 3.2 so I can't compare it to the earlier version t tried.

Just my two-cents worth....



-------- Original message --------
Subject: Re: [ampache/ampache] search.js (#2130)

**Comment by AshotN** [comment]
The idea is the implement as much of the client as possible. All the settings management can become API based so a new client can control it all. We can do a transition with the new client being opt-in and eventually opt-our if it’s good enough. 

I think one of the major flaws in Ampache, from what I’ve read on Reddit from users, is the interface is not intuitive, ancient and slow. 

The current web interface is really annoying to work on as it’s using PHP to fill in content, it becomes hard to follow what’s happening. The CSS is hard to edit just because CSS isn’t to easy to organize for big projects. A new interface can solve these problems since API based is easier to organize. Data in from API, and then React creates the DOM. 

I’m using Stylus to make the CSS manageable, TypeScript for type management on the client side. 

While I understand that the project is PHP based so it may be non ideal for some devs to have to also write React JavaScript. I believe that React is easier to maintain then PHP for a project of this size.

**Comment by wagnered** [comment]
Thanks for the explanation.

According to "Why Ampache" in the wiki, "By default, you can use the web interface which requires nothing (not proprietory e.w.)".

I think the web interface was meant to be a feature of Ampache.  Sure, make it more intuitive, but don't eliminate it.

I'm for having a "matching" client for Ampache.  If the existing API is adequately organized, I suggest you build additional code on that for your client to access.

As to the question of speed, according to the Ampache design philosophy: "Ampache takes pride in the fact that it is relatively fast..." and "Whenever possible, steps should be taken to increase the speed of the existing code."

With respect to your anecdotal evidence, slow compared to what? Plex? If there are a significant number of complaints, has performance testing been done and results compared with other streaming platforms using similar features?

The current web interface is hard to follow because it's not built on a framework. Javascript is strewn all throughout the various folders, etc.  Of course lack of documentation including comments in the code doesn't help.  A framework not only forces organization, a framework often has many tools built in and don't need to be reinvented.

(The following is another attempt to sell Laravel as the framework for Ampache) ( ;'>

For example, Laravel comes with Mix which simplifies building a JavaScript/react/vue front end. It can even compile and version Stylus and other stylesheets.

Using the Blade template engine, views can be inherited reducing the need for so many.  Also w3.css is designed for ln-line use and fewer .CSS files are needed.

Because Laravel has many out of the box features such as authorization, authentication, and the Blade templating engine which combines Laravel directives, php and HTTP, I recommend it as the framework of choice.  There are way too many features to list here.




-------- Original message --------
Subject: Re: [ampache/ampache] search.js (#2130)

**Comment by lachlan-00** [comment]
I do prefer not eliminating the web ui entirely. i like and have stuck with Ampache because of that simplicity (Apache+PHP+mariadb is just dead simple.)

I also think @AshotN is on the right path with extending things to allow alternatives. Ampache as a server and a tool is where it really shines and i'm fully supportive of continuing these efforts 100%

The code itself has come a long way this year too. large reductions in LOC and thousands of comment/docstrings. As for laravel... maybe one day @wagnered ;) i'm actually not averse to doing it but i still think i can keep this code going.

**Comment by AshotN** [comment]
I feel there might be confusion as to what I mean by client. When I'm referring to the web client, or client-side, I do mean the web interface. This isn't a standalone app, it's a part of Ampache. 

Ampache is slow compared to any modern website. The current client uses Ajax to avoid re-rendering the page, which is really ancient way of client side rendering. I have not taken any measurements to compare other clients. The server works fine, it's the client that is holding back Ampache. 

The code unmanageable in it's current state, it's a mess of unused CSS, scattered JavaScript where it's hard to follow what's happening. The current dependencies are outdated #2113 (for example)

I loaded a random test and there are 51 JavaScript files that get downloaded, and 16 CSS. three.js alone is 401 KB. While three.js is a really cool project, why does a music player need a JavaScript 3D library. jQuery files take about 800kb in combined CSS/JS files. 

With all these files the current client doesn't support keyboard shortcuts, like space to play/pause. 

No one ever stopped to clean this client, with everyone taking on a feature and a new dependency the client has hit a unsustainable point. Not to mention it wasn't written that well to begin with. CSS IDs are used instead of Classes. I recently made a pull request #2133, to cleanup some of the CSS. 

There is no minification or uglifier for the code. There is so much happening on the client right now it's a nightmare to fix it. 

And what is the point of fixing a client that looks like it's 10 years old? It needs to be redesigned. Also if the new client ever materializes, the old one doesn't have to disappear. It can be kept as the default, or as an opt-in. 

In regards to Laravel, I would love if the project was written in Laravel. But frankly that's a huge undertaking. I'm not versed enough in Laravel to do it. 

Also if Ampache ever does aspire to have a desktop client(I had actually worked on [one](https://github.com/AshotN/Ampact/), two years ago, which is why I made a JSON API) using the same React code we can make an Electron client pretty quickly.

**Comment by wagnered** [comment]
For a lack of a way to express it i' ll write: I've only heard the browser access to any back end referred to as an "interface" because the presentation doesn't exist separate from the back end and can be thought of as a control panel which would effectively be an extension of the back end.

Thanks for clearing that up, but I'm not so sure that it's entirely caused by the interface.  As I  understand the process, PHP out of the box isn't capable of threading so any JavaScript executing needs to wait for PHP to respond with the requested data to update the interface (respond to the JavaScript/jquery request.  If the php code isn't efficient, it will be the "bottleneck" that slows the refreshing of the display.  I don't believe that the php code is as efficient as it could be.

Something is required to access the XMLHttpRequest object to make the asynchronous calls:  If not Ajax, then what do you prefer?



-------- Original message --------
Subject: Re: [ampache/ampache] search.js (#2130)

**Comment by AshotN** [comment]
Well in the case of React the presentation exists separately from the server in a way. There is no server rendering in the interface I'm making. 

Generally, I hear people refer to the server responding to the client. Following the MVC model, the  database is the model, React is the client/view, and server is the Controller,  

The API is quite quick, I haven't noticed any delays in calls unless I ask to dump every song into JSON/XML. But that can be addressed pretty easily.

Ajax was really the only choice back then. I think fetching data from the server at an API is cleaner, easier to write, easier to manage, easier to fix. 

You're welcome to try out the client, I'm still working on it. But it does play music and search, so the bare bone functionality is there.

I understand the worry that Ampache will lose functionality, and that's not my intention. While I believe there is a lot of bloat that could be trimmed, I will add whatever the project maintainers believe should continue to exist.

**Comment by wagnered** [comment]
MVC is just a software architecture but some frameworks allow some requests and responses to be pipelined: this is a feature of Laravel. I think the following is how most people see the process:

 The controller must send the initial view to be displayed. Later the user at the interface (browser) causes React to send a jquery, Ajax or window.fetch request to the controller. The controller then returns just new data to the interface and React places the data on the display. Much faster than sending a new view.

So "there is no server rendering..." Does the initial view consist of only script which on being loaded into the browser causes it to request the actual view and data for rendering? Interesting,   I can envision that process but I don't quite understand how your process is more efficient and overall faster than the "normal" way to render a view with Ajax or window.fetch. It still has to render the entire view at least once. 

Or is the view transferred as data and script which is then rendered by JavaScript when the script is loaded into the browser and run?  I read only the overview on reactjs.org so I don't know what React is capable of doing.



-------- Original message --------
Subject: Re: [ampache/ampache] search.js (#2130)

**Comment by AshotN** [comment]
So React is responsible for generating a virtual DOM as it get’s data. So things like loading indicators are easier to build because the page is loaded very quickly. 

React Router handles client side routing.

React also makes handling to changes to a page simple. If let’s say the song changes, once the state(variables) change, React automatically renders the changes. 

Everything is broken into components for code reusability and faster renders. Views can pass data down to components, with props. Components can have components. 

I will try to make some instructions on how to compile the new interface. I’ll also provide a pre-compiled one if you don’t want to install yarn/node

**Comment by wagnered** [comment]
I would very much like to try it.  The concept is very interesting, but right now, the only access I have to the internet is through this old broken android phone.

-------- Original message --------
Subject: Re: [ampache/ampache] search.js (#2130)

**Comment by AshotN** [comment]
Whoever is brave enough to try to run the new interface. I added a README https://github.com/AshotN/ampache/tree/ReactClient

@lachlan-00 if you could try it and tell me if my instructions actually work, I'd appreciate it.

**Comment by kuzi-moto** [comment]
> The current web interface is hard to follow because it's not built on a framework.

I think having a framework may help, but honestly there are design patterns that Ampache could have followed from the start, such as separation of PHP and HTML, as AshotN has already mentioned relying on reusable classes than IDs, plus Ampache has a ton of dependencies already and I would like to see those reduced.

I really like that Ampache comes with a web interface, and I believe that no matter what, it should still come with a built-in interface. But the current code is really tough to work with, and is not built at all in a way that would make it easily mobile-friendly. I'm in the process of writing a new theme from scratch using only CSS, hence so many HTML changes I have proposed.

It seems like the web interface was designed specifically for Reborn, making it tough to implement a new theme. Maybe we don't need to scrap it completely, but it definitely needs some fixing up before a replacement can be made.

**Comment by AshotN** [comment]
@kuzi-moto I made a comment somewhere above clarifying that by client I meant an interface that ships with Ampache. Not one that is separate from the server. I apologize for the confusion. 

My fork has instructions to build my proposed interface. If you wish to try it out and see what how it feels. So you can get an idea of what we can get done.

**Comment by lachlan-00** [comment]
i'll give it a whirl this weekend
